polygon/sync/sync.go

```diff
@@ -144,6 +123,36 @@ func (s *Sync) onMilestoneEvent(
 	return nil
 }
 
+func (s *Sync) onMilestoneEvent(
+	ctx context.Context,
+	event EventNewMilestone,
+	ccBuilder CanonicalChainBuilder,
+) error {
+	milestone := event
+	if milestone.EndBlock().Uint64() <= ccBuilder.Root().Number.Uint64() {
+		return nil
+	}
+
+	milestoneHeaders := ccBuilder.HeadersInRange(milestone.StartBlock().Uint64(), milestone.Length())
+	err := s.milestoneVerifier(milestone, milestoneHeaders)
+	if err != nil {
+		s.logger.Debug(
+			syncLogPrefix("onMilestoneEvent: local chain tip does not match the milestone, unwinding to the previous verified milestone"),
+			"err", err,
+		)
+
+		if err := s.handleMilestoneMismatch(ctx, ccBuilder); err != nil {
```

Comment:
I think we may need to return here (even when no err), so can just do `return s.s.handleMilestoneMismatch`